### PR TITLE
Reorganize project files & reduce public interface

### PR DESCRIPTION
**Pull request by rebello95**
Reorganizes files in the project to be more clearly and consistently organized by visibility (whether the types within each file should be internal to the package or public). Also reduces the public interface of the module by moving some types to `internal` or `package`.

Note that `package` is specific to SPM, so the keyword cannot be used in conjunction with CocoaPods. For this reason, a few of the types are guarded with `#if COCOAPODS`. It's also worth noting that `ConnectNIO` cannot be consumed via CocoaPods today because NIO does not support CocoaPods. For this reason, the shared gRPC interfaces between `Connect` and `ConnectNIO` don't need to be exposed to CocoaPods at all. Additionally, [`package` requires Swift 5.9](https://github.com/apple/swift-evolution/blob/main/proposals/0386-package-access-modifier.md), so the Swift package version has been updated as well.

New directory structure:

```
.
├── Examples
│   ├── ElizaCocoaPodsApp
│   ├── ElizaSharedSources
│   │   ├── AppSources
│   │   └── GeneratedSources
│   │       └── connectrpc
│   │           └── eliza
│   │               └── v1
│   └── ElizaSwiftPackageApp
│       └── ElizaSwiftPackageApp
├── Libraries
│   ├── Connect
│   │   ├── Internal
│   │   │   ├── Generated
│   │   │   │   └── grpc
│   │   │   │       └── status
│   │   │   │           └── v1
│   │   │   ├── Interceptors
│   │   │   ├── Locks
│   │   │   ├── Streaming
│   │   │   └── Unary
│   │   ├── PackageInternal
│   │   ├── Public
│   │   │   ├── Implementation
│   │   │   │   ├── Clients
│   │   │   │   ├── Codecs
│   │   │   │   └── Compression
│   │   │   └── Interfaces
│   │   │       ├── Interceptors
│   │   │       └── Streaming
│   │   │           ├── AsyncAwait
│   │   │           └── Callbacks
│   │   └── proto
│   │       └── grpc
│   │           └── status
│   │               └── v1
│   ├── ConnectMocks
│   └── ConnectNIO
│       ├── Internal
│       │   └── Extensions
│       └── Public
├── Plugins
│   ├── ConnectMocksPlugin
│   ├── ConnectPluginUtilities
│   └── ConnectSwiftPlugin
└── Tests
    ├── ConnectLibraryTests
    │   ├── ConnectConformance
    │   ├── ConnectMocksTests
    │   ├── ConnectTests
    │   ├── Generated
    │   │   ├── connectrpc
    │   │   │   └── conformance
    │   │   │       └── v1
    │   │   └── server
    │   │       └── v1
    │   └── TestResources
    └── ConnectPluginUtilitiesTests
```

### CODE REVIEW

File: Libraries/ConnectNIO/Internal/GRPCInterceptor.swift [hunk 1]
```diff
@@ -186,3 +187,18 @@ extension GRPCInterceptor: StreamInterceptor {
         }
     }
 }
+
+private final class Locked<T>: @unchecked Sendable {
```
Comment (rebello95) [inline Libraries/ConnectNIO/Internal/GRPCInterceptor.swift:191]: Re-implements `Locked` using NIO's locks so that the `Connect.Locked` can stay internal

**Review by eseay — APPROVED** [review]

File: Package.swift [hunk 2]
```diff
@@ -75,6 +77,9 @@ let package = Package(
                 "buf.work.yaml",
                 "proto",
                 "README.md",
+            ],
+            swiftSettings: [
+                .unsafeFlags(["-package-name", packageName])
```
Comment (eseay) [inline Package.swift:82]: Have we confirmed that this is not going to cause issues? I have heard and experienced at some point in the past that using unsafe flags causes issues with downstream consumers of the dependency.

**Review by eseay — CHANGES_REQUESTED** [review]
Actually - quick question before approval.

File: Package.swift (hunk 2 above)
Comment (rebello95) [reply Package.swift:82]: Given the Eliza SwiftPM app builds fine with the `Connect` package I think it's okay? Do you know what else we should check? I'm not familiar with those issues

Comment (eseay) [reply Package.swift:82]: @rebello95 could we try archiving the Eliza app as a Release IPA? I think that would tell us.

Comment (rebello95) [reply Package.swift:82]: Yep, archiving seemed to work fine:
Screenshot 2023-11-13 at 3 45 27 PM

**Review by eseay — APPROVED** [review]

File: Package.swift (hunk 2 above)
Comment (rebello95) [reply Package.swift:82]: Seems you were right 🙃 #217

**Pull request merged**